clang/lib/CIR/Dialect/Transforms/LifetimeCheck.cpp

```diff
@@ -1291,16 +1427,68 @@ void LifetimeCheckPass::checkStore(StoreOp storeOp) {
     return;
   }
 
+  // Check if storing a moved-from value
+  auto data = storeOp.getValue();
+  if (auto loadOp = data.getDefiningOp<cir::LoadOp>()) {
+    auto srcAddr = loadOp.getAddr();
+    if (getPmap().count(srcAddr) &&
+        getPmap()[srcAddr].count(State::getInvalid()) &&
+        !owners.count(srcAddr) && !ptrs.count(srcAddr)) {
+      checkPointerDeref(srcAddr, storeOp.getLoc());
+    }
+
+    // Check if this is initialization from an rvalue (std::move)
+    // This handles cases like: int b(std::move(a));
+    auto destAddr = storeOp.getAddr();
+
+    // Only check for new variables (AllocaOp) that are Value types
+    if (auto allocaOp = destAddr.getDefiningOp<cir::AllocaOp>()) {
+      if (!owners.count(destAddr) && !ptrs.count(destAddr) &&
+          getPmap().count(srcAddr) && !owners.count(srcAddr) &&
+          !ptrs.count(srcAddr)) {
+        // Check if source is in a valid state (not moved-from or invalid)
+        // If srcAddr is NOT moved-from and we're doing a non-deref load,
+        // this could be an rvalue initialization
+        if (!getPmap()[srcAddr].count(State::getInvalid()) &&
+            !loadOp.getIsDeref()) {
+          // Mark source as moved-from for rvalue initialization
+          markPsetInvalid(srcAddr, InvalidStyle::MovedFrom, storeOp.getLoc());
+        }
+      }
+    }
+  }
+
+  // Handle coroutine tasks BEFORE the Value type reinitialization check
+  // Tasks need special tracking for dangling references
+  if (!ptrs.count(addr)) {
+    if (currScope->localTempTasks.count(storeOp.getValue())) {
+      checkCoroTaskStore(storeOp);
+      return;
+    }
+  }
+
+  // Handle reinitialization of Value types
+  if (getPmap().count(addr) && !owners.count(addr) && !ptrs.count(addr)) {
```

Comment:
Is reinitialization handling vital to have most of the basic tests working? Looks like this should come in another follow up PR